packages/jsii-kernel/lib/serialization.ts

```diff
@@ -296,16 +296,29 @@ export const SERIALIZERS: {[k: string]: Serializer} = {
         throw new Error(`Expected object reference, got ${JSON.stringify(value)}`);
       }
 
-      // Similarly to other end, we might be getting a reference type where we're
-      // expecting a value type. Accept this for now.
+      const namedType = host.lookupType((optionalValue.type as spec.NamedTypeReference).fqn);
+      const props = propertiesOf(namedType, host.lookupType);
+
+      if (Array.isArray(value)) {
```

Comment:
Wouldn't it make more sense to verify that the type is the expected type instead of what it is *not*. What if it's a map for example or a primitive?

Reply:
A map is actually correct, that's what a struct is (or should be),  and we do already check that a primitive is not passed.

We just missed the Array because of the fact that `typeof array === 'object'` in JavaScript.

By doing the explicit check we also get a nicer error message.